### PR TITLE
Fix backward for SDPA NT jagged layout

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115576

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -3452,65 +3451,103 @@ def test_sdpa(self, device, dtype):
...
+        check_forward_backward()
 
         # Test dispatcher works by calling only mem-effn and math (as they are safe for all devices)
-        with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=True, enable_math=False):
-            attn_nt = torch.nn.functional.scaled_dot_product_attention(q_nt, k_nt, v_nt).transpose(1, 2)
-
-            attn_nts = attn_nt.unbind()
-            self.assertEqual(attn_d1, attn_nts[0].unsqueeze(0), atol=output_ref_atol, rtol=output_ref_rtol)
-            self.assertEqual(attn_d2, attn_nts[1].unsqueeze(0), atol=output_ref_atol, rtol=output_ref_rtol)
+        with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=False, enable_math=False):
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:3538]: why are all 3 backends excluded here?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 2]
```diff
@@ -293,14 +293,15 @@ _scaled_dot_product_efficient_attention_nestedtensor_cuda(
...
+  int64_t max_seqlen_batch_k{0};
   std::tie(
       query_buffer_reshaped,
       key_buffer_reshaped,
       value_buffer_reshaped,
       cumulative_sequence_length_q,
       cumulative_sequence_length_kv,
       max_seqlen_batch_q,
-      std::ignore,
+      max_seqlen_batch_k,
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:304]: nice, we were already calculating this

**Review by jbschlosser — APPROVED** [review]
Thanks for the fix!

File: test/test_nestedtensor.py (hunk 1 above)
Comment (soulitzer) [reply test/test_nestedtensor.py:3538]: whoops typo, good catch!

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 3]
```diff
@@ -1165,7 +1167,8 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, c10::SymInt, c10::SymInt> _efficient_
       std::move(seed_t),
       std::move(offset_t),
       max_seqlen_q,
-      max_seqlen_k);
+      // TODO: why isn't this being set in the kernel?
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1170]: can you maybe make in issue or add more detail on this comment and what we expect

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -14506,7 +14506,7 @@
     CUDA: _flash_attention_backward
 
 # Returns ouput, logsumexp if compute_logsumexp
-- func: _efficient_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int? max_seqlen_q, float dropout_p, int custom_mask_type, bool compute_log_sumexp=False, *, float? scale=None, Tensor? causal_diagonal=None, Tensor? seqlen_k=None) -> (Tensor output, Tensor logsumexp, Tensor philox_seed, Tensor philox_offset, SymInt max_seqlen_batch_q, SymInt max_seqlen_batch_k)
+- func: _efficient_attention_forward(Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int? max_seqlen_q, int? max_seqlen_k, float dropout_p, int custom_mask_type, bool compute_log_sumexp=False, *, float? scale=None, Tensor? causal_diagonal=None, Tensor? seqlen_k=None) -> (Tensor output, Tensor logsumexp, Tensor philox_seed, Tensor philox_offset, SymInt max_seqlen_batch_q, SymInt max_seqlen_batch_k)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:14509]: if the meta registration is typed can you update

File: test/test_nestedtensor.py [hunk 5]
```diff
@@ -3452,65 +3451,103 @@ def test_sdpa(self, device, dtype):
...
+        k_nt = key(x_nt).view(*x_nt.size()[0:2], n_heads, head_dims).detach().requires_grad_(True)
+        k_nt_t = k_nt.transpose(1, 2)
+        v_nt = value(x_nt).view(*x_nt.size()[0:2], n_heads, head_dims).detach().requires_grad_(True)
+        v_nt_t = v_nt.transpose(1, 2)
+
+        attn_d2 = torch.nn.functional.scaled_dot_product_attention(q_d2_t, k_d2_t, v_d2_t).transpose(1, 2)
+        d1_grads = torch.autograd.grad(attn_d1.sum(), (q_d1, k_d1, v_d1))
+        d2_grads = torch.autograd.grad(attn_d2.sum(), (q_d2, k_d2, v_d2))
+
+        def check_forward_backward():
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nestedtensor.py:3521]: Oh we aren't using gradcheck here, we're just comparing the gradients between NT and non-NT. 
(And we don't actually support math kernel today for jagged layout NT!)

**Review by ani300 — APPROVED** [review]
Lgtm! The test that is failing might be because I was passing some extra stuff through the `output_nt_info` on top of thr offsets
